web/ui/react-app/src/pages/graph/CMExpressionInput.tsx

```diff
@@ -0,0 +1,242 @@
+
+import { EditorView, highlightSpecialChars, keymap, ViewUpdate, placeholder } from '@codemirror/view';
+import { EditorState, Prec, Compartment } from '@codemirror/state';
+import { indentOnInput, syntaxTree } from '@codemirror/language';
+import { history, historyKeymap } from '@codemirror/history';
+import { defaultKeymap, insertNewlineAndIndent } from '@codemirror/commands';
+import { bracketMatching } from '@codemirror/matchbrackets';
+import { closeBrackets, closeBracketsKeymap } from '@codemirror/closebrackets';
+import { searchKeymap, highlightSelectionMatches } from '@codemirror/search';
+import { commentKeymap } from '@codemirror/comment';
+import { lintKeymap } from '@codemirror/lint';
+import { PromQLExtension } from 'codemirror-promql';
+import { autocompletion, completionKeymap, CompletionContext, CompletionResult } from '@codemirror/autocomplete';
+import { theme, promqlHighlighter } from './CMTheme';
+
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { faSearch, faSpinner, faGlobeEurope } from '@fortawesome/free-solid-svg-icons';
+import MetricsExplorer from './MetricsExplorer';
+import { CompleteStrategy, newCompleteStrategy } from 'codemirror-promql/complete';
+import { computeStartCompletePosition } from 'codemirror-promql/complete/hybrid';
+
+const promqlExtension = new PromQLExtension();
+
+interface CMExpressionInputProps {
+  value: string;
+  onExpressionChange: (expr: string) => void;
+  queryHistory: string[];
+  metricNames: string[];
+  executeQuery: () => void;
+  loading: boolean;
+  enableAutocomplete: boolean;
+  enableHighlighting: boolean;
+  enableLinter: boolean;
+}
+
+const dynamicConfigCompartment = new Compartment();
+
+// Autocompletion strategy that wraps the main one and enriches
+// it with past query items.
+export class HistoryCompleteStrategy implements CompleteStrategy {
+  private complete: CompleteStrategy;
+  private queryHistory: string[];
+  constructor(complete: CompleteStrategy, queryHistory: string[]) {
+    this.complete = complete;
+    this.queryHistory = queryHistory;
+  }
+
+  promQL(context: CompletionContext): Promise<CompletionResult | null> | CompletionResult | null {
+    const result: Promise<CompletionResult | null> = Promise.resolve(this.complete.promQL(context));
+
+    const { state, pos } = context;
+    const tree = syntaxTree(state).resolve(pos, -1);
+    const start = computeStartCompletePosition(tree, pos);
+
+    // We are not completing at the very beginning, so only return normal matches.
+    if (start !== 0) {
+      return result;
+    }
+
+    return result.then(res => {
+      const historyItems: CompletionResult = {
+        from: start,
+        to: pos,
+        options: this.queryHistory.map(q => ({
+          label: q.length < 80 ? q : q.slice(0, 76).concat('...'),
+          detail: 'past query',
+          apply: q,
+          info: q,
+        })),
+        span: /^[a-zA-Z0-9_:]+$/,
```

Comment:
Yeah, I wasn't sure about that either. But if I remove the span here completely, then the completion dropdown disappears and reappears on each typed character (which is slow and ugly), because the search is starting completely from zero.

Reply:
maybe you can do that then:

```typescript
span: res !== nul ? res.span : /^[a-zA-Z0-9_:]+$/,
```
Like that you won't erase what is coming from the previous complete object

Reply:
or let's keep it like you did. And let's see if there is some weird cases ^^. I don't really think someone will be able to reach this case anyway. So should be good